code_search/code_search/pipeline.py

```diff
@@ -0,0 +1,91 @@
+    google_cloud_options.project = args.project
+    google_cloud_options.job_name = args.job_name
+    google_cloud_options.temp_location = '{}/{}/temp'.format(args.storage_bucket, args.job_name)
+    google_cloud_options.staging_location = '{}/{}/staging'.format(args.storage_bucket, args.job_name)
+
+    options.view_as(SetupOptions).setup_file = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'setup.py')
+
+    return options
+
+
+class SplitRepoPath(beam.DoFn):
+    def process(self, element, *args, **kwargs):
+        nwo, path = element.pop('repo_path').split(' ')
+        element['nwo'] = nwo
+        element['path'] = path
+        yield element
+
+
+class TokenizeCodeDocstring(beam.DoFn):
+    def process(self, element, *args, **kwargs):
+        from code_search.tokenizer import get_function_docstring_pairs
+        element['pairs'] = get_function_docstring_pairs(element.pop('content'))
+        yield element
+
+
+class ExtractFuncInfo(beam.DoFn):
+    def __init__(self, info_keys):
+        super(ExtractFuncInfo, self).__init__()
+
+        self.info_keys = info_keys
+
+    def process(self, element, *args, **kwargs):
+        info_rows = map(lambda pair: dict(zip(self.info_keys, pair)), element.pop('pairs'))
+        info_rows = map(lambda info_dict: self.merge_two_dicts(info_dict, element), info_rows)
+        yield info_rows
+
+    @staticmethod
+    def merge_two_dicts(dict_a, dict_b):
+        result = dict_a.copy()
+        result.update(dict_b)
+        return result
+
+
+class BigQueryGithubFiles(beam.PTransform):
+    def __init__(self, project, query_string, output_string):
+        super(BigQueryGithubFiles, self).__init__()
+
+        self.project = project
+        self.query_string = query_string
+        self.output_dataset, self.output_table = output_string.split(':')
+
+        self.data_columns = ['nwo', 'path', 'function_name', 'lineno', 'original_function',
+                             'function_tokens', 'docstring_tokens']
+        self.data_types = ['STRING', 'STRING', 'STRING', 'INTEGER', 'STRING', 'STRING', 'STRING']
+
+    def expand(self, pcoll):
+        return (pcoll
+                 | "Read BigQuery Rows" >> io.Read(io.BigQuerySource(query=self.query_string, use_standard_sql=True))
+                 | "Split 'repo_path'" >> beam.ParDo(SplitRepoPath())
+                 | "Tokenize Code/Docstring Pairs" >> beam.ParDo(TokenizeCodeDocstring())
+                 | "Extract Function Info" >> beam.ParDo(ExtractFuncInfo(self.data_columns[2:]))
+                 | "Flatten Rows" >> beam.FlatMap(lambda x: x)
+                 | "Write to BigQuery" >> io.WriteToBigQuery(project=self.project,
+                                                             dataset=self.output_dataset,
+                                                             table=self.output_table,
+                                                             schema=self.create_output_schema())
+                )
```

Comment:
I'm super intrigued by whatever is going on in this `expand` function.  Looks really useful way of somehow piping data transformations and really interesting use of the `|` and `>>` operators.  Would love to learn more about this if someone has time to explain!  Thanks :)

Reply:
Hey @hamelsmu, this is an Apache Beam pipeline. Everything between `|` and `>>` is just a manual description of the transform (by default it takes in the function signature next to the pipe). This part only helps with a clean labelling of the generated DAG. Anything after that are transformations, formally `PTransforms` and `DoFn`s. Without the syntactic cruft, they just take in stateless functions to perform transformations with the argument being the element from the previous step in the pipeline.

Reply:
Thanks for the info @activatedgeek I will definitely 👀 into this for my workflows from now on, looks really cool!